### ISSUE TITLE
Nested Namespaces and their names

### ISSUE THREAD

**Issue by dietzc** [imagej/imagej-ops#355]
Hi all,

@ctrueden is currently working on "nested namespace". I think now it would be a good time to discuss our `Namespace`s a bit more in detail, i.e. which `Namespace` do we want and how to we want to nest them. Currently, our `Namespace` names are somehow mixed: `ImageNamespace` describes on what kind of input it operates, however for example the `FilterNamespace`  describes more what it does. 

Current `Namespaces`:

* `ConvertNamespace`
* `CopyNamespace`
* `CreateNamespace`
* `DeconvolveNamespace`
* `FilterNamespace`
* `GeomNamespace`
* `HaralickNamespace`
* `ZernikeNamespace`
* `ImageMomentsNamespace`
* `ImageNamespace`
* `LabelingNamespace`
* `LBPNamespace`
* `LogicNamespace`
* `MathNamespace`
* `StatsNamespace`
* `TamuraNamespace`
* `ThreadNamespace`
* `Thresholdnamespace`

Here is one suggestions to start the discussion (I added some `Namespace`s):

* `Math`
** `FFT`
** `Logic`
** `Statistics`
** `Arithmetic`

* `Processing`
** `Thresholding`
** `Filters`
** `Morphological`
** `Convolution`
** `Deconvolution`

* `Features`
** `Pixel`
** `Object`
*** `Intensity`
**** `Haralick`, `Tamura', `ImageMoments`, `Stats`, `Zernike`, `LBP`.
*** `Geometric`

* `Transforms`
** `Convert`
** `Affine`
** `Dimensionality` (e.g. `Crop`, `RemoveDim`, etc).

* `Create`
** `Labeling`
** `Image`
** `Kernel`
** `Copy` 

* `Mappings`

* `Segmentation`

* `Threads` (?)
 

I've no idea if this is a good suggestion, but maybe it's good to get started.

Christian

**Comment by ctrueden** [comment]
Even though I think we need to support nested namespaces, I still want to avoid excessive nesting/categorization. So putting e.g. `logic` into `math` seems unnecessarily verbose.

Here is an example of how I envision script code looking soon:
```javascript
// @Ops ops
// @Ops.Create create
// @Ops.Create.Kernel kernel

myImg = create.img()
// also OK, but longer:
myImg = ops.create.img()

// for nested namespaces:
gaussKernel = kernel.gauss(...)
// also OK, but longer:
gaussKernel = create.kernel.gauss(...)
// and longer still:
gaussKernel = ops.create.kernel.gauss(...)
```

So while it will be possible to write `@Ops.Math.Logic logic` to achieve `logic.and(...)` usage in a succinct way, I think it will also be convenient to declare only `@Ops ops` and then `ops.math.logic.and(...)`, which I find inferior to `ops.logic.and(...)` here.

In other words: I am OK with quite a few top-level namespaces, in the interest of keeping common operations as short as possible. The nesting should only be used in cases where we are basically doing it anyway already, such as `create.kernelGauss` vs. `create.kernelLog` etc.

**Comment by dietzc** [comment]
`Features` are certainly a candidate for nesting, too. Also I would love to get rid of the `ImageNamespace` in favour of a categorization which describe what the `Op`s do, rather their input-type.

**Comment by ctrueden** [comment]
Agreed about `ImageNamespace`. I thought we already made this list somewhere...? At least partially here: https://github.com/imagej/imagej-ops/pull/168#issuecomment-172928009

**Comment by LeonYang5114** [comment]
Can we have the deep nested namespaces but at the same time provide shortcuts for unambiguous namespace calls? For example (using the scheme proposed by @dietzc ):
```
// this always works
ops.feature.object.intensity.haralic.correlation(...);
// type safe (not really sure) way that always works
ops.ns(HaralicNamespace.class).correlation(...);
// type unsafe way that works as long as the haralic namespace is unique
ops.ns("haralic").correlation(...);
```
I think the second way will work. The user just need to import the class, which could be done automatically in Eclipse. But if you are talking about scripting, probably the third way will be good enough?
For the type unsafe way, for now we could just use the bottom most namespace name (`haralic` in the example above), but we can also do something like `object.haralic` (even skipping the middle `intensity` namespace) to provide hints for identifying namespaces when the namespace system gets complicated. So it will search `haralic` namespace under all namespaces that are named `object`.

**Comment by bnorthan** [comment]
For reference here is how [scikit-image](http://scikit-image.org/docs/stable/api/api.html) and [matlab image processing](http://scikit-image.org/docs/stable/api/api.html) are organized.

Will Ops eventually have color space transformations??  Looks like that is missing right now.

**Comment by ctrueden** [comment]
Here's a question for y'all: do you foresee any "name clashes" in the nesting? E.g. `math.logic` and `stats.logic` as two different nested namespaces?

If not, we could simply expose _all_ leaf namespaces directly under `Ops`, _as well as_ nested...
